core/src/main/java/org/apache/iceberg/rest/RESTSessionCatalog.java

```diff
@@ -0,0 +1,626 @@
+      this.token = token;
+      this.tokenType = tokenType;
+    }
+
+    public Map<String, String> headers() {
+      return headers;
+    }
+
+    public String token() {
+      return token;
+    }
+
+    public String tokenType() {
+      return tokenType;
+    }
+
+    public AuthSession refresh() {
+      Map<String, String> request = OAuth2Util.tokenExchangeRequest(
+          token, tokenType, ImmutableList.of(RESTCatalogProperties.CATALOG_SCOPE));
+
+      OAuthTokenResponse response = client.postForm(
+          ResourcePaths.token(), request, OAuthTokenResponse.class, headers(), ErrorHandlers.defaultErrorHandler());
+
+      this.token = response.token();
+      this.tokenType = response.issuedTokenType();
+      this.headers = merge(headers, authHeaders(token));
+
+      if (response.expiresInSeconds() != null) {
+        scheduleTokenRefresh(this, response.expiresInSeconds());
+      }
+
+      return this;
+    }
+  }
+
+  private static Map<String, String> merge(Map<String, String> target, Map<String, String> updates) {
+    ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
+
+    target.forEach((key, value) -> {
+      if (!updates.containsKey(key)) {
+        builder.put(key, value);
+      }
+    });
+
+    updates.forEach(builder::put);
+
+    return builder.build();
+  }
+
+  private static Map<String, String> authHeaders(String token) {
+    if (token != null) {
+      return ImmutableMap.of(RESTCatalogProperties.AUTHORIZATION_HEADER, RESTCatalogProperties.BEARER_PREFIX + token);
+    } else {
+      return ImmutableMap.of();
+    }
+  }
+
+  private static Map<String, String> configHeaders(Map<String, String> properties) {
+    return RESTUtil.extractPrefixMap(properties, RESTCatalogProperties.HEADER_PREFIX);
+  }
+
+  private static ConfigResponse fetchConfig(RESTClient client, Map<String, String> headers) {
+    ConfigResponse configResponse = client
+        .get(ResourcePaths.config(), ConfigResponse.class, headers, ErrorHandlers.defaultErrorHandler());
+    configResponse.validate();
+    return configResponse;
+  }
+
+  private static OAuthTokenResponse exchangeToken(RESTClient client, Map<String, String> headers,
+                                                  String subjectToken, String subjectTokenType,
+                                                  String actorToken, String actorTokenType, String  scope) {
+    Map<String, String> request = OAuth2Util.tokenExchangeRequest(
+        subjectToken, subjectTokenType, actorToken, actorTokenType,
+        scope != null ? ImmutableList.of(scope) : ImmutableList.of());
+
+    OAuthTokenResponse response = client.postForm(
+        ResourcePaths.token(), request, OAuthTokenResponse.class, headers, ErrorHandlers.defaultErrorHandler());
+    response.validate();
+
+    return response;
+  }
+
+  public static OAuthTokenResponse fetchClientCredentials(RESTClient client, Map<String, String> headers,
```

Comment:
nit: naming is actually backwards here.  We're actually fetching a token (not credentials).

Reply:
I understand that this is the client credential flow, but it feels awkward to read (maybe `fetchClientCredentialFlow`?)

Reply:
Renamed to `fetchToken`.